orttraining/orttraining/training_ops/cuda/nn/batch_norm_internal.cc

```diff
@@ -0,0 +1,157 @@
+  Tensor* saved_inv_std = p_op_kernel_context->Output(4, channel_shape);
+
+  auto x_data = reinterpret_cast<const CudaT*>(X->template Data<T>());
+  auto scale_data = reinterpret_cast<const CudaT1*>(scale->template Data<T1>());
+  auto b_data = reinterpret_cast<const CudaT1*>(B->template Data<T1>());
+  auto mean_data = reinterpret_cast<const CudaT2*>(mean->template Data<T2>());
+  auto var_data = reinterpret_cast<const CudaT2*>(var->template Data<T2>());
+
+  auto y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T>());
+
+  const auto alpha = Consts<CudaT>::One;
+  const auto beta = Consts<CudaT>::Zero;
+
+  CudnnTensor data_desc, bn_tensor_desc;
+  vector<int64_t> new_dims;
+  BatchNormHelper::NormalizeDims(x_shape, new_dims);
+  ORT_RETURN_IF_ERROR(data_desc.Set(new_dims, CudnnTensor::GetDataType<CudaT>()));
+  // for fp16 input, `bn_tensor_desc` will have a float type; otherwise it will be the same as input type.
+  ORT_RETURN_IF_ERROR(bn_tensor_desc.Set(data_desc, cudnn_batch_norm_mode_));
+
+  auto running_mean_data = reinterpret_cast<CudaT2*>(running_mean->template MutableData<T2>());
+  auto running_var_data = reinterpret_cast<CudaT2*>(running_var->template MutableData<T2>());
+  auto saved_mean_data = reinterpret_cast<CudaT2*>(saved_mean->template MutableData<T2>());
+  auto saved_inv_std_data = reinterpret_cast<CudaT2*>(saved_inv_std->template MutableData<T2>());
+
+  auto p_scale = reinterpret_cast<const void*>(scale_data);
+  auto p_B = reinterpret_cast<const void*>(b_data);
+  auto p_running_mean = reinterpret_cast<void*>(running_mean_data);
+  auto p_running_var = reinterpret_cast<void*>(running_var_data);
+  auto p_saved_mean = reinterpret_cast<void*>(saved_mean_data);
+  auto p_saved_inv_std = reinterpret_cast<void*>(saved_inv_std_data);
+
+
+  const int64_t C = new_dims[1];
+  IAllocatorUniquePtr<float> p_f_scale, p_f_B, p_f_running_mean, p_f_running_var, p_f_saved_mean, p_f_saved_inv_std;
+
+  if (std::is_same<T1, MLFloat16>::value) {
+    // Convert scale/B to float
+    p_f_scale = GetScratchBuffer<float>(C);
+    p_f_B = GetScratchBuffer<float>(C);
+
+    Impl_Cast<CudaT1, float>(Stream(), scale_data, p_f_scale.get(), C);
+    Impl_Cast<CudaT1, float>(Stream(), b_data, p_f_B.get(), C);
+
+    p_scale = p_f_scale.get();
+    p_B = p_f_B.get();
+  }
+
+  if (std::is_same<T2, MLFloat16>::value) {
+    // Convert mean/var to float
+    p_f_running_mean = GetScratchBuffer<float>(C);
+    p_f_running_var = GetScratchBuffer<float>(C);
+    p_f_saved_mean = GetScratchBuffer<float>(C);
+    p_f_saved_inv_std = GetScratchBuffer<float>(C);
+
+    Impl_Cast<CudaT2, float>(Stream(), mean_data, p_f_running_mean.get(), C);
+    Impl_Cast<CudaT2, float>(Stream(), var_data, p_f_running_var.get(), C);
+
+    p_running_mean = p_f_running_mean.get();
+    p_running_var = p_f_running_var.get();
+    p_saved_mean = p_f_saved_mean.get();
+    p_saved_inv_std = p_f_saved_inv_std.get();
+  } else if (mean_data != running_mean_data) {
```

Comment:
I might have missed something... When will this case happen ?

Reply:
When if the alias output is the graph output. The allocation_planer will first check if graph output, if yes, then the allocation type will set to kAllocate. The Alias flag is checked after that. It's also reasonable that the input is an intermediate tensor but the output is graph output, we cannot use the input tensor directly as graph output.

Reply:
ic.